Add example of parameter estimation from multiple measurement trials

Peter230655 opened the pull request:
In a simple spring / damper model the spring constant and the dampening coefficient are to be estimated using noisy measurements.
The model is set up four times corresponding to four noisy measurements.
I create these noisy measurements by integrating the four systems and adding random numbers to each point of the solutions and a random shift.
As the method plot_trajectories does not work if no input trajectories are given, I plot them 'manually'.
(I hope I understood the idea correctly)

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,227 @@
+    u1.subs({t: t0}) - x0[4],
+    u2.subs({t: t0}) - x0[5],
+    u3.subs({t: t0}) - x0[6],
+    u4.subs({t: t0}) - x0[7],
```
You generally don't want instance constraints in parameter estimation problems. This overly constrains the solution.

Peter230655 replied:
> You generally don't want instance constraints in parameter estimation problems. This overly constrains the solution.

Sorry, I missed this in the PR I just started. Will correct it.

moorepants replied:
My bet is that if you remove these, you'll get better parameter estimates.

moorepants replied:
>  Question: Would you not know the initial conditionsd of the experiment, like im my case you would know how much you extended the mas, and at what initial speed it started?

You know the initial conditions only as good as your error filled measurement might say. If you force the initial condition to the value of a measurement that has error, then you are not allowing for solutions that minimize the overall error.

Peter230655 replied:
Makes sense!
When I give reasonable bounds it finds good values.
I will try to incorporate all you suggested tomorrow.

Peter230655 replied:
Question: does it make sense to estimate continuous parameters this way? What I mean: say, I apply a time varying force to this sytem and try to estimate what it looks like?

moorepants replied:
If I understand you, I think you can.

moorepants commented:
This is the right idea and very cool to see that it works!

Peter230655 commented:
> This is the right idea and very cool to see that it works!

Removing the instance control did two things:
- it made the estimates worse (I know their true value :-) )
- it gives a mistake in plot_constraint_violations
I will fix the error in the method over the weekend and fix it.
I think, no much merging this simulation until the method plot_constraint_violation is fixed?

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,205 @@
+are to be estimated.
+
+The idea is to set up four sets of eoms, one for each of the measurements, with
+identical parameters, and let opty estimate the parameters.
```
Maybe we should add an explanation of why we have to do such a thing. Something like:

For parameter identification, it is common to collect measurements of a system's trajectories from distinct expeirments. For example, if you are identifying the parameters of a mass-spring-damper system you will excite the system with different initial conditions multiple times. The data cannot simply be stacked and the identification run because the measurement data would be discontinuous between trials. A work around in opty is to create a set of differential equations with unique state variables for each measurement trial that all share the same constant parameters. You can then identify the parameters from all measurement trials simultaneously by passing the uncoupled differential equations to opty.

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,205 @@
+- :math:`x_1`: position of the mass of the first system [m]
+- :math:`x_2`: position of the mass of the second system [m]
+- :math:`x_3`: position of the mass of the third system [m]
+- :math:`x_4`: position of the mass of the fourth system [m]
```
These are really all the same system. I would call $x_1$ the state variable associated with the first measurement trial.

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,205 @@
+q_ind = [x1, x2, x3, x4]
+u_ind = [u1, u2, u3, u4]
+
+KM = me.KanesMethod(N, q_ind, u_ind, kd_eqs=kd)
```
I think the example would be clearer if you didn't use KanesMethod. You can write these EoMs directly with SymPy and it is readable.

```python
Matrix([
    x1.diff(t) - u1,
    m*u1.diff(t) + c*u1 + k*(x1 - l0),
])
```

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,205 @@
+fr, frstar = KM.kanes_equations(bodies, forces)
+eom = kd.col_join(fr + frstar)
+sm.pprint(eom)
+
```
make a new cell here and print the eoms

Peter230655 replied:
> make a new cell here and print the eoms

I will incorporate the changes.
What I do not fully understand: why not use Kane's method? After all I will print the eoms.

moorepants replied:
If you can write the equations directly in 8 lines of code, there is no need to obscure that behind all the kanesmethod stuff. The purpose of KanesMethod is to give you the equations of motion, but if we already know the equations of motion, why use it?

moorepants replied:
For example the EoM of a single penudulum is dead simple, so I just write them here:

https://opty.readthedocs.io/stable/examples/plot_pendulum_swing_up_fixed_duration.html#sphx-glr-examples-plot-pendulum-swing-up-fixed-duration-py

and a reader doesn't have to know anything about KanesMethod to understand that example. Not all systems need an advanced multibody dynamcis tool to find the EoMs.

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,205 @@
+ax.set_title('Measurements of the position of the mass')
+ax.legend();
+
+plt.show()
```
show is only needed once at the very end

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,205 @@
+#
+# If some measurement is considered more reliable, its weight w can be increased.
+#
+# objective = :math:`\int_{t_0}^{t_f} (w_1 (x_1 - noisy_{x_1})^2 + w_2 (x_2 - noisy_{x_2})^2 + w_3 (x_3 - noisy_{x_3})^2 + w_4 (x_4 - noisy_{x_4})^2)\, dt`
```
Using $x_1 - x_1^m$ where "m" is for measured would make these equations simpler

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,205 @@
+# %%
+# Initial guess.
+#
+initial_guess = np.array(list(np.random.randn(8*num_nodes + 2)))
```
In parameter identification problems you have the measured state values, so there is really no reason to not use the measurements as the initial guess for the state trajectories.

moorepants replied:
You also typically have some reasonable guess for the parameters and the parameters needs to be bounded.

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,205 @@
+# Solve the Optimization Problem.
+#
+solution, info = problem.solve(initial_guess)
+print(info['status_msg'])
```
new cell before the plot command, basically a new cell should be after anything you print or plot

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,205 @@
+x0 = np.array([3, 3, 3, 3, 0, 0, 0, 0])
+pL_vals = [1.0, 0.25, 1.0, 1.0]
+
+resultat1 = solve_ivp(gradient, t_span, x0, t_eval = times, args=(pL_vals,))
```
A more realistic measurement data set would be collected from trials that have different initial conditions. You often don't have control of the initial conditions in the experiments, you simply measure whatever the state happens to be.

moorepants commented:
I removed instance constraints, added bounds to the unknown parameters, set the initial guess of x to the measured states and set the initial guess for u's to zeros, and gave reasonable guesses for the unknown parameters and I get a consistent optimal solution. I think making those updates would make this more realistic and solve well. Additionally, we should use different initial conditions for each of the four measurement trials.

Peter230655 commented:
> I removed instance constraints, added bounds to the unknown parameters, set the initial guess of x to the measured states and set the initial guess for u's to zeros, and gave reasonable guesses for the unknown parameters and I get a consistent optimal solution. I think making those updates would make this more realistic and solve well. Additionally, we should use different initial conditions for each of the four measurement trials.

I will do this. But could you plot the constraint violations?

moorepants commented:
It would be helpful to make the final plot be a comparison of the optimal trajectories vs the measured trajectories (like done in the other parameter id examples). You can do this by first plotting the measurements on the same number of axes that the `plot_trajectories()` method produces and then pass that axes set into the plot method.

moorepants commented:
> I will do this. But could you plot the constraint violations?

No there is a bug.

moorepants commented:
There should be a `Problem.num_instance_constraints` that returns 0 when there are none.

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,205 @@
+# %%
+"""
+
+Parameter Estimation
```
The title and filename are too generic.

1. We use the terms "parameter identification" in opty not "parameter estimation", we should be consistent.
2. This is a special parameter id example, in that it is showing how to use multiple measurement trials that constitute a set of discontinuous data. So the title should reflect this uniqueness (relative to the other simpler parameter id examples)

Peter230655 commented:
I just pushed this to let you know where I am. Not yet been able to incorporate all your suggestions.
NB: I assumed only the positions of the mass are measured, not the speeds. So the measurement has only four plots, the plot_trajectories has 8.
(I have a hell of a time setting up eoms without sympy mechanics! Got the signs wrong all the time!)

Peter230655 commented:
> > I will do this. But could you plot the constraint violations?
> 
> No there is a bug.

I will try to fix the bug in the next few days. I might have caused it when I changed this method a bit.

moorepants commented:
Already fixed: https://github.com/csu-hmc/opty/pull/232

moorepants commented:
>  (I have a hell of a time setting up eoms without sympy mechanics! Got the signs wrong all the time!)

Equations are here: https://en.wikipedia.org/wiki/Mass-spring-damper_model

Peter230655 commented:
> > (I have a hell of a time setting up eoms without sympy mechanics! Got the signs wrong all the time!)
> 
> Equations are here: https://en.wikipedia.org/wiki/Mass-spring-damper_model

Thanks!
I studied engineering 100 years ago! I should still be able to do it without Wikipedia! :-(

Peter230655 commented:
> Already fixed: #232

Unbelievable! I would have taken me 2 days!! :-(

moorepants commented on examples-gallery/plot_parameter_estimation.py:
```diff
@@ -0,0 +1,209 @@
+"""
+
+Parameter Identification of a Mass-Spring-Damper System.
+========================================================
```
I'd prefer "Parameter Identification from Non-Continuous Measurements" of something similar. The unique aspect of this example relative to what we already have is that we have these independent measurements of the system's motion.

moorepants replied:
Maybe "non-contiguous" is the right word.

moorepants replied:
Parameter Identification from Non-Contiguous Measurements

Peter230655 replied:
> I'd prefer "Parameter Identification from Non-Continuous Measurements" of something similar. The unique aspect of this example relative to what we already have is that we have these independent measurements of the system's motion.

Will do so and push

moorepants replied:
Also change the filename to `plot_non_contiguous_parameter_estimation.py`

moorepants commented on examples-gallery/plot_non_continuous_parameter_estimation.py:
```diff
@@ -0,0 +1,209 @@
+# %%
+"""
+
+Parameter Identification from Non-Continuous Measurements.
```
You didn't agree that "contiguous" was the correct word?

contiguous /kən-tĭg′yoo͞-əs/
adjective

    Sharing an edge or boundary; touching.
    Neighboring; adjacent.
    Connecting without a break.

moorepants replied:
I guess it can be either, but we have continuous measurements, they are just not contiguous in time.

Peter230655 replied:
> You didn't agree that "contiguous" was the correct word?
> 
> contiguous /kən-tĭg′yoo͞-əs/ adjective
> 
> ```
> Sharing an edge or boundary; touching.
> Neighboring; adjacent.
> Connecting without a break. 
> ```

:-)  :-)
I thought is was a typo. Knowing you I should have known better! Will change and push.

moorepants replied:
It doesn't' really matter, I just get nit picky on wording. I'm working an journal article this morning doing the final editing and my mind is in the mode of being extremely precise with wording.

Peter230655 replied:
> It doesn't' really matter, I just get nit picky on wording. I'm working an journal article this morning doing the final editing and my mind is in the mode of being extremely precise with wording.

Can I get a copy when you are done with it?

moorepants replied:
Sure.

Peter230655 replied:
Thanks!

Peter230655 commented:
I modified this example a bit: I added a force = F1*sin(omegat*t) and also estimate F1 and omega.
Also I made the number of measurements variable.
Would you prefer this 'enlarged' simulation?
If yes, would I simply push to this PR, or close this PR and make a new one?
Thanks!

moorepants commented:
> I modified this example a bit: I added a force = F1 * sin(omega * t) and also estimate F1 and omega.

I'd prefer if we didn't add this to this example and simply demonstrate how to do parameter identification from multiple non-contiguous measurements. If you want it to be something different I can make an example that focus on the parameter id approach.

Peter230655 commented:
> > I modified this example a bit: I added a force = F1 * sin(omega * t) and also estimate F1 and omega.
> 
> I'd prefer if we didn't add this to this example and simply demonstrate how to do parameter identification from multiple non-contiguous measurements. If you want it to be something different I can make an example that focuses on the parameter id approach.

**NO**, not at all, we leave it as it is!
I play around a lot, the come up with something else, which I feel is interesting, so I let you know.
I do have a tendency to complicate my programs, I want to see how far I can go - and this has a tendency to cover up what should be shown.
NB: I do not know, why it failed on Github test.

moorepants commented:
>  NB: I do not know, why it failed on Github test.

You have to read the error message.

Peter230655 commented:
> > NB: I do not know, why it failed on Github test.
> 
> You have to read the error message.
Of course I tried to read it - problem is understanding :-(
![image](https://github.com/user-attachments/assets/51addb96-dd4e-49db-86c0-a591008603b8)

moorepants commented on examples-gallery/plot_non_contiguous_parameter_estimation.py:
```diff
@@ -0,0 +1,209 @@
+
+# %%
+# Set up the Identification Problem.
+# --------------------------------
```
The error is here and it is exactly what it says in the error message.

Peter230655 replied:
> The error is here and it is exactly what it says in the error message.

Fair enough! I only looked at the 'big' title. I will correct and push
